### PR TITLE
VS Code Java Cannot Run from Read-Only Location

### LINKED ISSUE #1301: VS Code Java Cannot Run from Read-Only Location
When running VS Code Java, some files are created in `server\config_win\org.eclipse.core.runtime` (this is on Windows, but likely not Windows-specific). 
The problem with this is that you cannot run VS Code Java from a location where you don't have write access. This prevents VS Code Java from running as a built-int extension in Theia, for example.
To me it seems that the files in question are created by the eclipse configuration management. IMO, those files should be created in the directory obtained from `ExtensionContext#globalStoragePath`. Eclipse has a startup parameter to set the osgi configuration area (see [here](https://help.eclipse.org/2019-12/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fmisc%2Fruntime-options.html&cp%3D2_1_5_0&anchor=osgiconfigurationarea) and [here](https://help.eclipse.org/2019-12/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fmisc%2Fruntime-options.html))

### PR DESCRIPTION
**Pull request by snjeza**
Fixes #1301 

Signed-off-by: Snjezana Peco <snjezana.peco@redhat.com>

### CODE REVIEW

File: src/javaServerStarter.ts [hunk 1]
```diff
@@ -114,11 +115,36 @@ function prepareParams(requirements: RequirementsData, javaConfiguration, worksp
...
+	let configuration = path.resolve(context.globalStoragePath, configDir);
+	ensureExists(configuration);
+	configuration = path.resolve(configuration, version);
+	ensureExists(configuration);
+	const configIniName = "config.ini";
+	const configIni = path.resolve(configuration, configIniName);
+	if (!fs.existsSync(configIni)) {
+		const ini = path.resolve(__dirname, '../server', configDir, configIniName);
+		fs.copyFileSync(ini, configIni);
+	}
```
Comment (testforstephen) [inline src/javaServerStarter.ts]: How about moving the logic to a method like resolveConfiguration()?

Comment (snjeza) [reply src/javaServerStarter.ts]: Done.

File: src/javaServerStarter.ts [hunk 2]
```diff
@@ -114,11 +115,41 @@ function prepareParams(requirements: RequirementsData, javaConfiguration, worksp
...
+	if (packageFile) {
+		version = packageFile.version;
+	}
+	else {
+		version = '0.0.0';
+	}
+	let configuration = path.resolve(context.globalStoragePath, configDir);
+	ensureExists(configuration);
+	configuration = path.resolve(configuration, version);
+	ensureExists(configuration);
```
Comment (fbricon) [inline src/javaServerStarter.ts]: it generates paths with version at the end: "/Users/fbricon/Library/Application Support/Code - Insiders/User/globalStorage/redhat.java/config_mac/0.58.0"

Makes more sense to have 0.58.0/config_mac

Comment (snjeza) [reply src/javaServerStarter.ts]: Fixed.

**Comment by snjeza** [comment]
@fbricon I have updated the PR.

**Comment by fbricon** [comment]
Thanks @snjeza!

**Comment by fbricon** [comment]
This changes causes makes dev mode unusable:
- if you rebuild a new version of jdt.ls with gulp build_server, the new jars are ignored. If you delete server/ before  calling gulp build_server, the extension fails to start  (server failed to start 5 times in the last 3 min). I believe this is caused by config.ini  being copied to the globalstate directory only once (the 1st time a new version of vscode-java is executed), and that file contains references to timestamped bundles:
```
#This configuration file was written by: org.eclipse.equinox.internal.frameworkadmin.equinox.EquinoxFwConfigFileParser
#Tue Mar 03 16:37:42 CET 2020
osgi.bundles=reference\:file\:com.google.gson_2.8.2.v20180104-1110.jar@4,reference\:file\:com.google.guava_27.1.0.v20190517-1946.jar@4,reference\:file\:com.ibm.icu_64.2.0.v20190507-1337.jar@4,reference\:file\:javax.inject_1.0.0.v20091030.jar@4,reference\:file\:javax.xml_1.3.4.v201005080400.jar@4,reference\:file\:org.apache.ant_1.10.7.v20190926-0324/@4,reference\:file\:org.apache.commons.lang3_3.1.0.v201403281430.jar@4,reference\:file\:org.apache.log4j_1.2.15.v201012070815.jar@4,reference\:file\:org.eclipse.ant.core_3.5.700.v20191221-1203.jar@4,reference\:file\:org.eclipse.buildship.core_3.1.4.v20200211-2258-s.jar@4,reference\:file\:org.eclipse.compare.core_3.6.800.v20191213-2246.jar@4,reference\:file\:org.eclipse.core.commands_3.9.700.v20191217-1850.jar@4,reference\:file\:org.eclipse.core.contenttype_3.7.600.v20200124-1609.jar@4,reference\:file\:org.eclipse.core.expressions_3.6.700.v20200212-1751.jar@4,reference\:file\:org.eclipse.core.filebuffers_3.6.900.v20200117-0616.jar@4,reference\:file\:org.eclipse.core.filesystem_1.7.700.v20200110-1734.jar@4,reference\:file\:org.eclipse.core.jobs_3.10.700.v20200106-1020.jar@4,reference\:file\:org.eclipse.core.net_1.3.800.v20200218-0937.jar@4,reference\:file\:org.eclipse.core.resources_3.13.700.v20200209-1624.jar@4,reference\:file\:org.eclipse.core.runtime_3.17.100.v20200203-0917.jar@4\:start,reference\:file\:org.eclipse.core.variables_3.4.800.v20200120-1101.jar@4,reference\:file\:org.eclipse.debug.core_3.15.0.v20200224-0654.jar@4,reference\:file\:org.eclipse.equinox.app_1.4.400.v20191212-0743.jar@4,reference\:file\:org.eclipse.equinox.common_3.11.0.v20200206-0817.jar@4,reference\:file\:org.eclipse.equinox.frameworkadmin_2.1.400.v20191002-0702.jar@4,reference\:file\:org.eclipse.equinox.frameworkadmin.equinox_1.1.300.v20200211-1503.jar@4,reference\:file\:org.eclipse.equinox.launcher_1.5.700.v20200207-2156.jar@4,reference\:file\:org.eclipse.equinox.launcher.cocoa.macosx.x86_64_1.1.1100.v20190907-0426/@4,reference\:file\:org.eclipse.equinox.preferences_3.7.700.v20191213-1901.jar@4,reference\:file\:org.eclipse.equinox.registry_3.8.700.v20200121-1457.jar@4,reference\:file\:org.eclipse.equinox.security_1.3.500.v20200114-1637.jar@4,reference\:file\:org.eclipse.equinox.security.macosx_1.101.200.v20190903-0934.jar@4,reference\:file\:org.eclipse.equinox.simpleconfigurator_1.3.500.v20200211-1505.jar@4,reference\:file\:org.eclipse.equinox.simpleconfigurator.manipulator_2.1.500.v20200211-1505.jar@4,reference\:file\:org.eclipse.jdt.apt.core_3.6.550.v20190919-0948.jar@4,reference\:file\:org.eclipse.jdt.apt.pluggable.core_1.2.400.v20190626-1334.jar@4,reference\:file\:org.eclipse.jdt.compiler.apt_1.3.800.v20200302-0636.jar@4,reference\:file\:org.eclipse.jdt.compiler.tool_1.2.800.v20200302-0636.jar@4,reference\:file\:org.eclipse.jdt.core_3.21.0.v20200302-0423.jar@4,reference\:file\:org.eclipse.jdt.core.manipulation_1.13.100.v20200224-2108.jar@4,reference\:file\:org.eclipse.jdt.debug_3.15.0.v20200224-0604/@4,reference\:file\:org.eclipse.jdt.junit.core_3.10.600.v20200214-0716.jar@4,reference\:file\:org.eclipse.jdt.junit.runtime_3.5.100.v20200214-0716.jar@4,reference\:file\:org.eclipse.jdt.launching_3.17.0.v20200215-0651.jar@4,reference\:file\:org.eclipse.jdt.launching.macosx_3.4.500.v20200116-2251.jar@4,reference\:file\:org.eclipse.jdt.ls.core_0.52.0.202003031536.jar@4\:start,reference\:file\:org.eclipse.lsp4j_0.10.0.v20200229-1029.jar@4,reference\:file\:org.eclipse.lsp4j.jsonrpc_0.10.0.v20200229-1029.jar@4,reference\:file\:org.eclipse.ltk.core.refactoring_3.10.400.v20200215-1156.jar@4,reference\:file\:org.eclipse.m2e.archetype.common_1.14.0.20191209-1923/@4,reference\:file\:org.eclipse.m2e.core_1.14.0.20191209-1925.jar@4,reference\:file\:org.eclipse.m2e.jdt_1.14.0.20191209-1925.jar@4,reference\:file\:org.eclipse.m2e.lifecyclemapping.defaults_1.14.0.20191209-1925.jar@4,reference\:file\:org.eclipse.m2e.maven.indexer_1.14.0.20191209-1923/@4,reference\:file\:org.eclipse.m2e.maven.runtime_1.14.0.20191209-1923/@4,reference\:file\:org.eclipse.m2e.maven.runtime.slf4j.simple_1.14.0.20191209-1923/@4,reference\:file\:org.eclipse.m2e.workspace.cli_0.3.1.jar@4,reference\:file\:org.eclipse.osgi.services_3.8.0.v20190206-2147.jar@4,reference\:file\:org.eclipse.osgi.util_3.5.300.v20190708-1141.jar@4,reference\:file\:org.eclipse.text_3.10.100.v20200217-1239.jar@4,reference\:file\:org.eclipse.xtend.lib_2.19.0.v20190902-0728.jar@4,reference\:file\:org.eclipse.xtend.lib.macro_2.19.0.v20190902-0728.jar@4,reference\:file\:org.eclipse.xtext.xbase.lib_2.19.0.v20190902-0728.jar@4,reference\:file\:org.gradle.toolingapi_6.0.0.v20200211-2258-s.jar@4,reference\:file\:org.hamcrest.core_1.3.0.v20180420-1519.jar@4,reference\:file\:org.jboss.tools.maven.apt.core_1.5.3.201911081053.jar@4,reference\:file\:org.junit_4.13.0.v20200204-1500.jar@4,reference\:file\:org.slf4j.api_1.7.2.v20121108-1250.jar@4
eclipse.application=org.eclipse.jdt.ls.core.id1
osgi.bundles.defaultStartLevel=4
osgi.framework=file\:plugins/org.eclipse.osgi_3.15.200.v20200214-1600.jar
eclipse.product=org.eclipse.jdt.ls.core.product
osgi.framework.extensions=reference\:file\:org.eclipse.osgi.compatibility.state_1.1.700.v20200207-2156.jar
eclipse.p2.data.area=@config.dir/../p2
eclipse.p2.profile=DefaultProfile
```
In this case org.eclipse.jdt.ls.core_0.52.0.202003031536.jar no longer exists

**Comment by testforstephen** [comment]
i also found two issues about the dev mode:

### Issue 1:
- Open VS Code on a normal Java project, and make sure the language server is activated.
- Open VS Code on vscode-java repo, rebuild a new version of jdt.ls with `gulp build_server`
- Launch an extension host via the debug configuration `Launch Extension` in launch.json, you will find the language server failed to start. 

The reason is the copy logic tries to clean up the existing config directory in the glob storage, and it throws uncaught exception because that directory is occupied by other language server process.

### Issue 2:
- Open VS Code on vscode-java repo, rebuild a new version of jdt.ls with `gulp build_server`
- Launch an extension host via the debug configuration `Launch Extension` in launch.json, and make sure the language server is activated in the new extension host window.
- Open a new VS Code on a normal Java project, you will get error saying language server fails to start.

The reason is the shared config.ini is replaced by the extension host. And because the jar version is different, this config.ini cannot be consumed by the plugins in the VS Code extension directory.

**Pull request merged**